### PR TITLE
🎉 New source: TiDB

### PR DESCRIPTION
**Pull request by Daemonxiao**
## What
Add new source [TiDB](https://en.pingcap.com/tidb/). 

## Recommended reading order
1. `docs/integrations/tidb.md`
2. `airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSource.java`
3. `/airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java`

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/SUMMARY.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

</details>

<details><summary><strong>Integration</strong></summary>

</details>

<details><summary><strong>Acceptance</strong></summary>

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-tidb/src/main/resources/spec.json [hunk 1]
```diff
@@ -0,0 +1,58 @@
+{
+  "documentationUrl": "https://docs.airbyte.io/integrations/sources/tidb",
+  "connectionSpecification": {
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Tidb Source Spec",
```
Comment (zhangyangyu) [inline airbyte-integrations/connectors/source-tidb/src/main/resources/spec.json:5]:
```suggestion
    "title": "TiDB Source Spec",
```

File: airbyte-integrations/connectors/source-tidb/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+## Testing
+We use `JUnit` for Java tests.
+
+### Unit and Integration Tests
+Place unit tests under `src/test/...`
+Place integration tests in `src/test-integration/...` 
+
+#### Acceptance Tests
+Airbyte has a standard test suite that all source connectors must pass. Implement the `TODO`s in
+`src/test-integration/java/io/airbyte/integrations/sources/tiDBSourceAcceptanceTest.java`.
```
Comment (Daemonxiao) [inline airbyte-integrations/connectors/source-tidb/README.md:48]:
```suggestion
`src/test-integration/java/io/airbyte/integrations/sources/TiDBSourceAcceptanceTest.java`.
```

File: docs/integrations/sources/tidb.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+| `double`                                  | number                 |                                                              |
+| `decimal`                                 | number                 |                                                              |
+| `binary`                                  | base64 binary string   |                                                              |
+| `blob`                                    | base64 binary string   |                                                              |
+| `date`                                    | string                 | ISO 8601 date string. ZERO-DATE value will be converted to NULL. If column is mandatory, convert to EPOCH. |
+| `datetime`, `timestamp`                   | string                 | ISO 8601 datetime string. ZERO-DATE value will be converted to NULL. If column is mandatory, convert to EPOCH. |
+| `time`                                    | string                 | ISO 8601 time string. Values are in range between 00:00:00 and 23:59:59. |
+| `year`                                    | year string            | [Doc](https://dev.mysql.com/doc/refman/8.0/en/year.html)     |
```
Comment (zhangyangyu) [inline docs/integrations/sources/tidb.md:99]:
```suggestion
| `year`                                    | year string            | [Doc](https://docs.pingcap.com/tidb/stable/data-type-date-and-time#year-type)     |

I suggest use TiDB doc link.
```

File: docs/integrations/sources/tidb.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+4. `SSH Connection Port` is the port on the bastion server with which to make the SSH connection. The default port for SSH connections is `22`, so unless you have explicitly changed something, go with the default.
+5. `SSH Login Username` is the username that Airbyte should use when connection to the bastion server. This is NOT the TiDB username.
+6. If you are using `Password Authentication`, then `SSH Login Username` should be set to the password of the User from the previous step. If you are using `SSH Key Authentication` TiDB password, but the password for the OS-user that Airbyte is using to perform commands on the bastion.
+7. If you are using `SSH Key Authentication`, then `SSH Private Key` should be set to the RSA Private Key that you are using to create the SSH connection. This should be the full contents of the key file starting with `-----BEGIN RSA PRIVATE KEY-----` and ending with `-----END RSA PRIVATE KEY-----`.
+
+### Data type mapping
+
+TiDB data types are mapped to the following data types when synchronizing data:
```
Comment (zhangyangyu) [inline docs/integrations/sources/tidb.md:78]:
```suggestion
[TiDB data types](https://docs.pingcap.com/tidb/stable/data-type-overview) are mapped to the following data types when synchronizing data:
```

**Comment by Daemonxiao** [comment]
refer issue #10891

**Comment by marcosmarxm** [comment]
Thanks for the contribution @Daemonxiao I'll review it tomorrow

**Comment by marcosmarxm** [comment]
@Daemonxiao is it not possible to use TiDB using the MySQL connector? The implementation looks is duplicating code from MySQL connector.

File: airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSource.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import io.airbyte.integrations.base.Source;
+import io.airbyte.integrations.base.ssh.SshWrappedSource;
+import io.airbyte.integrations.source.jdbc.AbstractJdbcSource;
+import java.util.List;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class TiDBSource extends AbstractJdbcSource<MysqlType> implements Source {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSource.java:22]: Please check the implementation of `mysql-strict-encryt` connector. So far I didn't find any change from Mysql that is specifically applied to TiDB.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Comment by zhangyangyu** [comment]
> @Daemonxiao is it not possible to use TiDB using the MySQL connector? The implementation looks is duplicating code from MySQL connector.

Thanks for your review @marcosmarxm. TiDB is a MySQL compatible database, just like CockroachDB vs postgresql. So in most cases using MySQL connector or even JDBC connector works. But the compatibility is not 100%. For example, TiDB has its own internal system databases, TiDB doesn't support spatial types. And the biggest incompatibility is TiDB doesn't support MySQL binlog at all so currently MySQL connector CDC would not function at all. And I expect in future the differences might vary. So we'd like to develop a dedicated TiDB source connector although currently it replicate many MySQL connector code.

**Comment by marcosmarxm** [comment]
> > @Daemonxiao is it not possible to use TiDB using the MySQL connector? The implementation looks is duplicating code from MySQL connector.
> 
> Thanks for your review @marcosmarxm. TiDB is a MySQL compatible database, just like CockroachDB vs postgresql. So in most cases using MySQL connector or even JDBC connector works. But the compatibility is not 100%. For example, TiDB has its own internal system databases(present in code), TiDB doesn't support spatial types(present in doc). And the biggest incompatibility is TiDB doesn't support MySQL binlog at all so currently MySQL connector CDC would not function at all(present in both code and doc). And I expect in future the differences might vary. So we'd like to develop a dedicated TiDB source connector although currently it replicate many MySQL connector code.

But it's possible to reduce the duplicate code right. The SQL operations can be the same because it's a subset of MySQL. You only need to create the spec and the building spec (see the connector I commented previously). Did you try to use only creating the TiDBSource.java file? Can you point in your code where is different from the MySQL implementation? Why not use the MySQL connector and only create a TiDB documentation explaining you can use the MySQL connector with ONLY the default method?

**Comment by zhangyangyu** [comment]
> But it's possible to reduce the duplicate code right. 

Definitely. We change the implementation from extending JDBC to MySQL, so no need to duplicate much code.

> The SQL operations can be the same because it's a subset of MySQL. You only need to create the spec and the building spec (see the connector I commented previously).

Simply extending the spec is not ideal since the difference is not only in configuration but also code behaviour. The difference is small for now but we also want to leave the possibility for future. For example, add CDC ability, directly accessing the storage layer of TiDB.

> Can you point in your code where is different from the MySQL implementation?

Please review the new implementation. :-)

File: airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import com.mysql.cj.result.Field;
+import io.airbyte.db.SourceOperations;
+import io.airbyte.integrations.source.mysql.MySqlSourceOperations;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+import static io.airbyte.db.jdbc.JdbcConstants.*;
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java:16]: Please convert to single import here.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSource.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  @Override
+  public Set<String> getExcludedInternalNameSpaces() {
+    return Set.of(
+        "information_schema",
+        "metrics_schema",
+        "performance_schema",
+        "mysql");
+  }
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSource.java:75]: The internal of TiDB is exactly as MySQL ones?

File: airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    } catch (final IllegalArgumentException ex) {
+      LOGGER.warn(String.format("Could not convert column: %s from table: %s.%s with type: %s (type name: %s). Casting to VARCHAR.",
+          field.get(INTERNAL_COLUMN_NAME),
+          field.get(INTERNAL_SCHEMA_NAME),
+          field.get(INTERNAL_TABLE_NAME),
+          field.get(INTERNAL_COLUMN_TYPE),
+          field.get(INTERNAL_COLUMN_TYPE_NAME)));
+      return MysqlType.VARCHAR;
+    }
+  }
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java:67]: You can remove this function for now, you're not overwritten it. If in the future will be a change let this happens with it, but today the behavior is equal to MySQL.

File: airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  public void setJsonField(ResultSet resultSet, int colIndex, ObjectNode json) throws SQLException {
+    final ResultSetMetaData metaData = (ResultSetMetaData) resultSet.getMetaData();
+    final Field field = metaData.getFields()[colIndex - 1];
+    final String columnName = field.getName();
+    final MysqlType columnType = field.getMysqlType();
+
+    switch (columnType) {
+      case BIT -> {
+        putBinary(json, columnName, resultSet, colIndex);
+      }
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java:32]: Can you link to TiDB docs showing BIT works only a binary?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
some comments

File: airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSource.java (hunk 7 above)
Comment (zhangyangyu) [reply airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSource.java:75]: No. TiDB is MySQL compatible but it's implemented from scratch, in golang and rust. So at least for these internal databases, it's definitely not 100% compatible.

File: airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java (hunk 8 above)
Comment (zhangyangyu) [reply airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java:67]: Not exactly same. `bit(1)` is not interpreted as boolean.

File: airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java (hunk 9 above)
Comment (zhangyangyu) [reply airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java:32]: The formal doc of BIT is https://docs.pingcap.com/tidb/stable/data-type-numeric#bit-type.

File: airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java (hunk 6 above)
Comment (Daemonxiao) [reply airbyte-integrations/connectors/source-tidb/src/main/java/io/airbyte/integrations/source/tidb/TiDBSourceOperations.java]: Thanks for you review. I've changed it.
[convert to single import](https://github.com/airbytehq/airbyte/pull/11283/commits/167fed07cd98db54e16beb7d778a2ead78d3818f)

**Comment by marcosmarxm** [comment]
Thanks @Daemonxiao I requested to the Connector team to give a final review of your contribution.

**Review by marcosmarxm — APPROVED** [review]
thanks @Daemonxiao

**Comment by Daemonxiao** [comment]
@marcosmarxm Thanks for your work. BTW, do we need to provide the TiDB logo for Airbyte UI?

**Comment by marcosmarxm** [comment]
> @marcosmarxm Thanks for your work. BTW, do we need to provide the TiDB logo for Airbyte UI?

you can add into airbyte-config/init/src/main/resources/icons
also update the seed file to receive the icon

**Pull request merged**